From ce9f5b0969700f2ec663312f7d16e3bcade6c583 Mon Sep 17 00:00:00 2001
From: David Drysdale <drysdale@google.com>
Date: Wed, 18 May 2016 13:46:40 +0100
Subject: [PATCH] Non-v3 cert not an error if not doing CAB checks

Also make it easier to add new messages in future.
---
 checks.c   | 9 ++++++++-
 checks.h   | 4 ++++
 messages.c | 7 ++++---
 3 files changed, 16 insertions(+), 4 deletions(-)

diff --git a/checks.c b/checks.c
index 8d871bc..301a7d1 100644
--- a/checks.c
+++ b/checks.c
@@ -1462,7 +1462,14 @@ void check(unsigned char *cert_buffer, size_t cert_len, CertFormat format, CertT
 	ret = X509_get_version(x509);
 	if (ret != 2)
 	{
-		SetError(ERR_NOT_VERSION3);
+		if (cab_checks)
+		{
+			SetError(ERR_NOT_VERSION3);
+		}
+		else
+		{
+			SetInfo(INF_NOT_VERSION3);
+		}
 	}
 	CheckASN1_integer(x509->cert_info->version);
 
diff --git a/checks.h b/checks.h
index b57fc13..28dc8a2 100644
--- a/checks.h
+++ b/checks.h
@@ -84,6 +84,7 @@ typedef enum { PEM, DER } CertFormat;
 #define ERR_EC_INVALID_GROUP_ORDER            73
 #define ERR_EC_INCORRECT_ORDER                74
 #define ERR_EC_NON_ALLOWED_CURVE              75
+#define ERR_LAST                              75
 
 /* This violates a SHOULD (or MUST with exception that can't be checked) */
 #define WARN_NON_PRINTABLE_STRING      0
@@ -98,6 +99,7 @@ typedef enum { PEM, DER } CertFormat;
 #define WARN_RSA_EXP_RANGE             9
 #define WARN_POLICY_QUALIFIER_NOT_CPS 10
 #define WARN_EXPLICIT_TEXT_ENCODING   11
+#define WARN_LAST                     11
 
 /* Certificate is valid, but contains things like deprecated or not checked. */
 #define INF_SUBJECT_CN                    0
@@ -105,6 +107,8 @@ typedef enum { PEM, DER } CertFormat;
 #define INF_CRL_NOT_URL                   2
 #define INF_UNKNOWN_VALIDATION            3        /* Software doesn't know OID yet. */
 #define INF_NAME_ENTRY_LENGTH_NOT_CHECKED 4        /* Software doesn't hnow how to check size yet. */
+#define INF_NOT_VERSION3                  5
+#define INF_LAST                          5
 
 extern bool cab_checks;
 extern uint32_t errors[];
diff --git a/messages.c b/messages.c
index dc30d0a..db45929 100644
--- a/messages.c
+++ b/messages.c
@@ -125,6 +125,7 @@ static const char *info_strings[] = {
 	"I: CRL is not a URL\n", /* INF_CRL_NOT_URL */
 	"I: Unknown validation policy\n", /* INF_UNKNOWN_VALIDATION */
 	"I: Name entry length not checked\n", /* INF_NAME_ENTRY_LENGTH_NOT_CHECKED */
+	"I: Certificate not version 3\n", /* INF_NOT_VERSION3 */
 };
 
 /* 
@@ -139,7 +140,7 @@ char *get_messages()
 	buffer = malloc(16384);
 	buffer[0] = '\0';
 
-	for (int i = 0; i <= ERR_EC_NON_ALLOWED_CURVE; i++)
+	for (int i = 0; i <= ERR_LAST; i++)
 	{
 		if (GetBit(errors, i))
 		{
@@ -147,7 +148,7 @@ char *get_messages()
 		}
 	}
 
-	for (int i = 0; i <= WARN_EXPLICIT_TEXT_ENCODING; i++)
+	for (int i = 0; i <= WARN_LAST; i++)
 	{
 		if (GetBit(warnings, i))
 		{
@@ -155,7 +156,7 @@ char *get_messages()
 		}
 	}
 
-	for (int i = 0; i <= INF_NAME_ENTRY_LENGTH_NOT_CHECKED; i++)
+	for (int i = 0; i <= INF_LAST; i++)
 	{
 		if (GetBit(info, i))
 		{
-- 
1.9.1

